src/uploader/uploader.cpp

```diff
@@ -0,0 +1,158 @@
+
+#include "util/make_unique.h"
+#include "util/configuration.h"
+#include "util/log.h"
+#include <iostream>
+#include <fstream>
+#include "services/httpservice.h"
+#include "services/httpparsing.h"
+#include "uploader/uploader.h"
+#include <boost/filesystem.hpp>
+
+UploadService::UploadService(std::streambuf *in, std::streambuf *out, std::streambuf *err)
+        : input(in), error(err), response(out)
+{
+
+}
+
+void UploadService::run(){
+    try {
+        //analog to HTTPService check these env variables
+        std::string request_method = getenv_str("REQUEST_METHOD", false);
+        std::string content_type = getenv_str("CONTENT_TYPE", true);
+
+        if (request_method != "POST") {
+            throw UploaderException("Uploader requires POST requests.");
+        }
+        if (content_type.find("multipart/form-data") != std::string::npos
+            && content_type.find("multipart/mixed") != std::string::npos)
+        {
+            throw UploaderException("Uploader requires a multipart POST requests.");
+        }
+
+        runInternal();
+    } catch (std::exception &e) {
+        response.send500(concat("Invalid upload: ", e.what()));
+    }
+}
+
+void UploadService::runInternal(){
+
+    MultipartReader mr(input);
+
+    if(!mr.hasNextPart()){
+        throw UploaderException("Empty multipart request.");
+    }
+
+    // parse first part as url parameters
+    Parameters params = parseParameters(mr);
+
+    // check user session
+    const std::string &sessiontoken = params.get("sessiontoken");
+    auto session = UserDB::loadSession(sessiontoken);
+    auto user = session->getUser();
+    auto username = user.getUsername();
+    if(!user.hasPermission("upload")){
+        throw OperatorException("User does not have permission to upload.");
+    }
+
+    std::string upload_dir = Configuration::get<std::string>("uploader.directory");
+    boost::filesystem::path total_path(upload_dir);
+    total_path /= username;
+    total_path /= params.get("upload_name");
```

Comment:
does boost take care of sanitizing the path? e.g. if the username contains "*" or "/"

Reply:
Better: use user *ID* instead. Usernames could potentially change over time

Reply:
I changed it to use the user ID. 
What would sanitizing mean for the upload_name or the filenames? Should I just reject the upload if a "/" is in a name? "*" are just kept as standard characters by boost.

Reply:
If boost can handle such file names it's ok. For invalid names we can return an error. Also: don't we get the original name of the file anyhow on upload using browser? Then normally there should be no invalid names when uploaded this way.